producer-c/producer-cloudwatch-integ/CanaryStreamCallbacks.cpp

```diff
@@ -0,0 +1,216 @@
+    pCanaryStreamCallbacks->streamCallbacks.freeStreamCallbacksFn = CanaryStreamFreeHandler;
+
+CleanUp:
+
+    if (STATUS_FAILED(retStatus)) {
+        pCanaryStreamCallbacks = NULL;
+    }
+
+    // Set the return value if it's not NULL
+    if (ppStreamCallbacks != NULL) {
+        *ppStreamCallbacks = (PStreamCallbacks) pCanaryStreamCallbacks;
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS freeCanaryStreamCallbacks(PStreamCallbacks* ppStreamCallbacks)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PCanaryStreamCallbacks pCanaryStreamCallbacks = NULL;
+
+    CHK(ppStreamCallbacks != NULL, STATUS_NULL_ARG);
+
+    pCanaryStreamCallbacks = (PCanaryStreamCallbacks) *ppStreamCallbacks;
+
+    // Call is idempotent
+    CHK(pCanaryStreamCallbacks != NULL, retStatus);
+
+    delete(pCanaryStreamCallbacks->timeOfNextKeyFrame);
+    // Release the object
+    MEMFREE(pCanaryStreamCallbacks);
+
+    // Set the pointer to NULL
+    *ppStreamCallbacks = NULL;
+
+CleanUp:
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS CanaryStreamFreeHandler(PUINT64 customData)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PStreamCallbacks pStreamCallbacks;
+
+    CHK(customData != NULL, STATUS_NULL_ARG);
+    pStreamCallbacks = (PStreamCallbacks) *customData;
+    CHK_STATUS(freeCanaryStreamCallbacks(&pStreamCallbacks));
+
+CleanUp:
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS CanaryStreamErrorReportHandler(UINT64 customData, STREAM_HANDLE streamHandle,
+                                               UPLOAD_HANDLE uploadHandle, UINT64 erroredTimecode,
+                                               STATUS statusCode)
+{
+    PCanaryStreamCallbacks pCanaryStreamCallbacks = (PCanaryStreamCallbacks) customData;
+
+    DLOGE("CanaryStreamErrorReportHandler got error %lu at time %" PRIu64 " for stream % " PRIu64 " for upload handle %" PRIu64, statusCode, erroredTimecode, streamHandle, uploadHandle);
+    pCanaryStreamCallbacks->streamErrorDatum.SetValue(1.0);
+    pCanaryStreamCallbacks->streamErrorDatum.SetUnit(Aws::CloudWatch::Model::StandardUnit::None);
+    CanaryStreamSendMetrics(pCanaryStreamCallbacks, pCanaryStreamCallbacks->streamErrorDatum);
+
+    return STATUS_SUCCESS;
+}
+
+STATUS CanaryStreamFragmentAckHandler(UINT64 customData,
+                                      STREAM_HANDLE streamHandle,
+                                      UPLOAD_HANDLE uploadHandle,
+                                      PFragmentAck ack)
+{
+    PCanaryStreamCallbacks pCanaryStreamCallbacks = (PCanaryStreamCallbacks) customData;
+    UINT64 time = GETTIME();
+    UINT64 timeOfFragmentEndSent = pCanaryStreamCallbacks->timeOfNextKeyFrame->find(ack->timestamp)->second;
+    if (ack->ackType != FRAGMENT_ACK_TYPE_BUFFERING) {
+        DLOGD("ack type %u latency %" PRIu64, ack->ackType, (GETTIME() - timeOfFragmentEndSent) / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
```

Comment:
You already have the time variable

Reply:
Agreed. Removed time variable and using GETTIME() everywhere